Tryouts/Plugins/ApplicationPlugins/MorganStanley.ComposeUI.ProcessExplorer/dotnet/src/ProcessExplorer.Abstractions/Processes/ProcessInfoMonitor.cs

```diff
@@ -0,0 +1,217 @@
+
+    public IObservable<KeyValuePair<int, ProcessStatus>> ProcessIds => _processIdsSubject;
+
+
+    public ProcessInfoMonitor(ILogger? logger)
+    {
+        _logger = logger ?? NullLogger.Instance;
+    }
+
+    public bool ContainsId(int processId)
+    {
+        lock (_processIdsLocker)
+        {
+            return _processIds.Contains(processId);
+        }
+    }
+
+    public void AddProcess(int processId)
+    {
+        lock (_processIdsLocker)
+        {
+            if (processId == 0) return;
+
+            _processIds.Add(processId);
+            _processIdsSubject.OnNext(new(processId, ProcessStatus.Running));
+        }
+    }
+
+    public void RemoveProcessId(int processId)
+    {
+        lock (_processIdsLocker)
+        {
+            var removedProcessId = _processIds.Remove(processId);
+            if (!removedProcessId) return;
+
+            _processIdsSubject.OnNext(new(processId, ProcessStatus.Terminated));
+            _logger.ProcessTerminatedInformation(processId);
+        }
+    }
+
+    public void SetProcessIds(
+        int mainProcessId,
+        ReadOnlySpan<int> processIds)
+    {
+        lock (_processIdsLocker)
+        {
+            foreach (var id in processIds)
+            {
+                try
+                {
+                    if (_processIds.Contains(id)) continue;
+
+                    var process = Process.GetProcessById(id);
+
+                    _processIds.Add(id);
+
+                    AddChildProcesses(id, process.ProcessName);
+                }
+                catch (Exception exception)
+                {
+                    _logger.ProcessExpected(exception);
+                }
+            }
+
+            if (mainProcessId != 0 && !_processIds.Contains(mainProcessId))
+                _processIds.Add(mainProcessId);
+        }
+    }
+
+    public ReadOnlySpan<int> GetProcessIds()
+    {
+        lock (_processIdsLocker)
+        {
+            return _processIds.ToArray();
+        }
+    }
+
+    public void ClearProcessIds()
+    {
+        lock (_processIdsLocker)
+        {
+            _processIds.Clear();
+        }
+    }
+
+    /// <summary>
+    /// Returns the PPID of the given process.
+    /// </summary>
+    /// <param name="processId"></param>
+    /// <returns></returns>
+    public abstract int? GetParentId(int processId, string processName);
+
+    /// <summary>
+    /// Returns the memory usage (%) of the given process.
+    /// </summary>
+    /// <returns></returns>
+    public abstract float GetMemoryUsage(int processId, string processName);
+
+    /// <summary>
+    /// Returns the CPU usage (%) of the given process.
+    /// </summary>
+    /// <returns></returns>
+    public abstract float GetCpuUsage(int processId, string processName);
+
+    /// <summary>
+    /// Continuously watching created processes.
+    /// </summary>
+    public virtual void WatchProcesses(int mainProcessId)
+    {
+        ClearProcessIds();
```

Comment:
I'd add a lock around this method as well, to avoid a scenario like:
thread1 ClearProcessIds;
thread1 AddProcess
thread2 ClearProcessIds
thread1AddChildProcesses;